development_guidelines.md

```diff
@@ -0,0 +1,424 @@
+  - All requirements are exhaustively tested by one or more tests.
+  - All specifications are linked against their corresponding tests 
+(how-to link see [here](#how-to-link-a-requirement-against-a-test))
+- User interface:
+  - The user interface is easy to understand and use.
+  - The user interface is comprehensively documented.
+  - If an existing user interface is broken by the new feature, then the
+interface version has to be adapted according the nature of the change.
+- System architecture:
+  - The system architecture complies with the clean architecture guidelines.
+(see "Clean Architecture" by Robert C. Martin)
+  - The architecture diagrams are up to date and easily understandable.
+- Code:
+  - The code has to comply with the clean code rules 
+(see "Clean Code: A Handbook of Agile Software Craftsmanship" by Robert C. Martin).
+- Tests:
+  - The tests are easy to understand (e.g. each test only checks one aspect 
+of a component, etc.)
+  - The tests are comprehensively documented (how-to document tests see
+[here](#how-to-document-tests))
+
+
+## Continuous integration (CI)
+All pull requests (PR's) have to be approved by the continuous integration tool
+[Travis](https://travis-ci.org/PilzDE/).  
+
+### How-to activate Travis for forked repositories
+If you are an external developer and want to create a pull request, you can
+do so by forking one of our repositories, make changes, and then create a pull
+request on our repository.  
+To check that the continuous integration runs without problems on 
+*your* forked repository, you have to activate Travis for your repository.
+For more information on how to do that have a look at the
+[Travis tutorials](https://docs.travis-ci.com/user/tutorial/#to-get-started-with-travis-ci).  
+  
+
+### How-to run the CI processes locally
+If you have a look at our
+[Travis file](https://github.com/PilzDE/pilz_robots/blob/melodic-devel/.travis.yml#L34), 
+you see that our CI is based on the 
+[ROS industrial CI configurations](https://github.com/ros-industrial/industrial_ci)
+for ROS. In some cases, for example in case of errors only appearing on 
+Travis, etc., it might be advantageous to do *exactly* what Travis does locally 
+(At least in terms of building the packages and executing the tests).
+To achieve this, have a look at the explanations given in the 
+[ros-industrial repository](https://github.com/ros-industrial/industrial_ci/blob/legacy/doc/index.rst#run-industrial_ci-on-local-host).
+
+## Code style guide
+We follow the [ROS coding style guide](http://wiki.ros.org/CppStyleGuide). 
+Currently, the styling of the code is not automatically checked by the CI, 
+so please make sure that you comply with the ROS coding style guide. 
+
+### How-to format code automatically
+If you want to automatically format your code using the ROS style guide,
+then see [here](http://wiki.ros.org/CppStyleGuide#Autoformatting_of_ROS_Code).
```

Comment:
Please avoid (here-links)[http://stephanieleary.com/2015/05/why-click-here-is-a-terrible-link-and-what-to-write-instead/] for easier readability. Thanks.

Reply:
Done.